### ISSUE TITLE
Portus not able to authenticate itself using portus username

### ISSUE THREAD

**Issue by aleksnikolic** [SUSE/Portus#376]
Hi @mssola . Opening a new issue as the #338 is now closed.

I am running the latest portus version from master that includes your yesterday's changes in log verbosity when Portus is not allowed to pull manifest from the registry.

```ruby
Could not fetch the tag for target {"mediaType"=>"application/vnd.docker.distribution.manifest.v1+json", "size"=>23172, "digest"=>"sha256:74d37aa2f8486bfc81bc3ca3e80978a4826034c4dd5804abc2090948c4fde7d8", "length"=>23172, "repository"=>"aleks/distribution", "url"=>"https://distribution.testcompany.co.uk/v2/aleks/distribution/manifests/sha256:74d37aa2f8486bfc81bc3ca3e80978a4826034c4dd5804abc2090948c4fde7d8"}
Reason: Cannot obtain authorization token: #<Net::HTTPForbidden:0x007fb08a722940>
Completed 202 Accepted in 167ms (ActiveRecord: 1.6ms)
```

I have checked the database and Portus user does exist:
```sql
mysql> select * from users where username = 'portus'\G;
*************************** 1. row ***************************
                    id: 1
              username: portus
                 email: portus@portus.com
    encrypted_password: $2a$10$P.SZmuTM84m2YIJeTKvBketwsq281r4hhynuh2pSbNswDmxCLaFpy
  reset_password_token: NULL
reset_password_sent_at: NULL
   remember_created_at: NULL
         sign_in_count: 1
    current_sign_in_at: 2015-09-29 15:35:35
       last_sign_in_at: 2015-09-29 15:35:35
    current_sign_in_ip: 172.17.0.2
       last_sign_in_ip: 172.17.0.2
            created_at: 2015-09-29 15:05:45
            updated_at: 2015-09-29 15:35:35
                 admin: 1
               enabled: 1
             ldap_name: NULL
       failed_attempts: 0
             locked_at: NULL
1 row in set (0.00 sec)
```
The password is correct as I was able to login through remote docker client using username 'portus' and the password define in secrets.yml and I was able to push images to registry. 

```bash
aleksandarnikolic~:$ docker login distribution.testcompany.co.uk
Username (aleks): portus
Password:
Email (aleks@testcompany.co.uk): portus@portus.com
WARNING: login credentials saved in /Users/aleksnikolic/.dockercfg.
Login Succeeded
aleksandarnikolic~:$
```

Is there any manual step that maybe I have missed during setup?  The reason why I haven't used your compose setup script is because I have mysqldb on the remote host. The only manual thing i did is:
```ruby
rake db:create db:migrate db:seed
```

The registry config is:
```bash
version: 0.1
log:
  fields:
    service: registry
storage:
    cache:
        layerinfo: inmemory
    filesystem:
        rootdirectory: /data/registry
http:
    addr: 0.0.0.0:5000
    debug:
       addr: 0.0.0.0:5001
auth:
  token:
    realm: https://portus.testcompany.co.uk/v2/token
    service: distribution.testcompany.co.uk
    issuer: portus.test.lan
    rootcertbundle: /registry/portus.crt
notifications:
  endpoints:
    - name: portus
      url: http://portus:3000/v2/webhooks/events
      timeout: 500ms
      threshold: 5
      backoff: 1s
```
My compose file is:
```bash
distribution:
     environment:
        -  VIRTUAL_HOST=distribution.testcompany.co.uk
        -  VIRTUAL_PORT=5000
        -  DOCKER_HOST=distribution2.testcompany.co.uk
        -  CLIENT_MAX_BODY_SIZE=0
     image: docker.testcompany.co.uk/distribution:latest
     volumes:
        -  "/data/registry:/data/registry"
     expose:
        -  5000
        -  5001
     links:
        - portus
portus:
     environment:
        -  VIRTUAL_HOST=portus.testcompany.co.uk
     image: docker.testcompany.co.uk/portus:latest
     external_links:
        -  mysqldambassador_ambassador_1:db
     expose:
        -  3000
     command: thin start
     volumes:
        -  "/usr/local/keys/database.yml:/portus/config/database.yml:ro"
```

Please let me know if you need more details.

**Comment by brycereitano** [comment]
I have ran into the same issue today running on the latest master branch. I was using the new portusctl setup method.

This is preventing Portus to synchronize with the registry.

**Comment by brycereitano** [comment]
I decided to test the generated password for the user 'portus' and then an user created account. Passwords left out for obvious reasons.

I have earlier tested Devise has a maximum password length of 72 characters. Anything past 72 characters are ignored completely. I'm pretty sure this is not the cause, but still a concern on false security. Source: http://security.stackexchange.com/questions/39849/does-bcrypt-have-a-maximum-password-length

```
irb(main):016:0> User.find_by(username: "canonical").valid_password?('securepassword')
  User Load (0.4ms)  SELECT  `users`.* FROM `users` WHERE `users`.`username` = 'canonical' LIMIT 1
=> true
irb(main):017:0> User.find_by(username: "portus").valid_password?('<some super long password generated earlier>')
  User Load (0.4ms)  SELECT  `users`.* FROM `users` WHERE `users`.`username` = 'portus' LIMIT 1
=> false
irb(main):018:0>
```

**Comment by brycereitano** [comment]
I apologize for the repeated postings, but I'm pretty confident on the issue.

If you run "portusctl setup" multiple times, it will create a new `@portus_password` each time, updating the secrets file. However, the database is already seeded, so it doesn't update the existing user with the new password.

The two ways I'm seeing on fixing this are:
- Keep track if you already generated a password and don't generate a new one. 
- Generate a new password every time, but update the existing portus user.

**Comment by flavio** [comment]
> If you run "portusctl setup" multiple times, it will create a new @portus_password each time, updating the secrets file. However, the database is already seeded, so it doesn't update the existing user with the new password.

You are right, I'm on the issue

**Comment by mssola** [comment]
@brycereitano your issue should be fixed with the commit cfac89c4b1de2f243d38c7b88d65c93ffc84f0af.

**Comment by flavio** [comment]
@brycereitano in the meantime you can fix your instance of Portus in that way:

```
portusctl exec rails c
# then type
User.where(username: 'portus').first.update!(password: Rails.application.secrets.portus_password)
```

**Comment by flavio** [comment]
Leaving the initial issue to @mssola

**Comment by brycereitano** [comment]
Thanks, @mssola and @flavio!

@aleksnikolic Does this fix your issue as well?

**Comment by aleksnikolic** [comment]
@brycereitano I have checked out the latest tag but the issue is still there. Any guidance how to trace it down?

**Comment by travisghansen** [comment]
@aleksnikolic are you running this with apache/passenger or using ```bundle exec thin start``` or similar?

I'm running into this using thin and I believe it's a concurrency issue.  For example *something* is triggering the:
```POST "/v2/webhooks/events"```
and when that happens I immediately try to access the portus web interface in my browser (which just sits there waiting).  After the waiting a few moments I get the error on the portus log:
```Reason: Cannot obtain authorization token: #<Net::HTTPGatewayTimeOut:0x0000000518d4b8>```

At which point my browser loads the page appropriately.  Subsequently another
```POST "/v2/webhooks/events"```
comes through at which point my browser interface blocks again.  Repeat.

Taking it a step further, since my browser interface is blocking I'm certain the docker client trying to grab an auth token simultaneously is doing the same thing.

Basically the app trying to handle 2 requests simultaneously is killing it.  In this particular case:
```POST "/v2/webhooks/events"```
which eventually triggers a reciprocal
```Started GET "/v2/token?account=portus&scope=...```

If anyone has any tips to overcome this issue I'd be happy to test.  I'm not really a ruby dev so I'm too unfamiliar with some of these components :(

**Comment by flavio** [comment]
@travisghansen yes this is a concurrency issue caused by thin. Take a look at [this](https://github.com/SUSE/Portus/wiki/An-NGinx-setup) section of our wiki for a working configuration of nginx and thin.

**Comment by travisghansen** [comment]
@flavio yeah, I'm running behind nginx already (and further up from that haproxy acutally).  I'm currently running with ```server 127.0.0.1:3000 fail_timeout=5s;``` in lieu of sockets.  However is it the multiple servers that really solves the issue?

**Comment by flavio** [comment]
@travisghansen yes it does

**Comment by travisghansen** [comment]
@flavio what's the relation between --servers and --max-conns in this context then?  Seems a bit strange to have a single request block the whole application.  Does this mean the number of concurrent requests is limited to the number of servers I start?

**Comment by travisghansen** [comment]
OK, I can confirm adding multiple servers resolves the issue here.  I'd still like to understand the question from the comment above however.  Thanks!

**Comment by flavio** [comment]
Thin is constrained by slow connections.

This is what is going on:
  * Something happens on your registry (like a push)
  * Registry sends a notification to Portus
  * Thin receives a POST request from the registry containing the event just occurred
  * Before replying with a 200, Portus fires a new request against the registry to have more details about that even (in the past that wasn't necessary because the event had all the information we needed)
  * To talk with the registry Portus has to authenticate itself
  * That leads to another authentication request against Portus (yes, that Portus authenticating against itself... well, there's no other choice)
  * However the thin worker is still stuck dealing with the 1st request, and that hangs the whole sytem

Even though the docker-compose setup is meant to be used just for development purposes, I think it's time to replace thin with something else (maybe puma)...

I recommend reading [this](http://www.nateberkopec.com/2015/07/29/scaling-ruby-apps-to-1000-rpm.html) article.

There's a relevant part about thin describing this sort of issues:

> Thin is multi-threaded, not multi-process, and threads only run one at a time on MRI. So while actually running your application, your host becomes unavailable (with all the negative consequences outlined under the Webrick section above). Unless you get very fancy with your use of EventMachine, too, Thin cannot accept other requests while waiting for I/O in the application code to finish. For example - if your application code POSTs to a payments service for credit card authorization, Thin cannot accept new requests while waiting for that I/O operation to complete by default. Essentially you'd need to modify your application code to send events back to Thin's EventMachine reactor loop to tell Thin "Hey, I'm waiting for I/O, go do something else". Here's more about how that works.
> 
> Thin can deal with slow client requests, but it can't deal with slow application responses or application I/O without a whole lot of custom coding.

**Comment by travisghansen** [comment]
@flavio yeah, that all sounds sane, great explanation.  As an FYI I'm *not* using docker-compose but a custom-made docker image based on centos 7 with nginx and supervisor (and some other garbage) built-in to the image as well.

As I mentioned I'm not really a ruby guy but unicorn seems to be what I see most often as the web-server in these kinds of projects.

Thanks!

**Comment by aleksnikolic** [comment]
Sorry guys, I finally got some time to try what you suggested.

@travisghansen I am using NGINX but as a separate docker container as a http router that routes the traffic on the same docker host towards registry and portus containers. 

After swapping thin with puma server the concurrency issue is not fully gone (I still see some lag and UI freezes for couple of seconds) but at least I have a working setup now and Portus auth against itself is not timing out anymore. I am running puma server with args ```-t 8:32 -w 3``` but that probably needs little more tuning.

**Comment by mssola** [comment]
Thanks @aleksnikolic for your feedback. What do you mean by lag and UI freezes ? What's the exact way to reproduce those? I'm going to close this issue, since it's been resolved, but I'd like to know about your issues, so feel free to comment :)

**Comment by aleksnikolic** [comment]
@mssola No need to reproduce. It was puma server misconfig. Just to confirm that the issue (wasn't the Portus issue in the first place at all) is fully gone now when I swapped the server to phusion passenger. I am using the official phusion/passenger-ruby22:0.9.17 image with nginx enabled.

Thanks again guys for your support and I hope you are going to carry on with the great work on Portus.

**Comment by mssola** [comment]
@aleksnikolic good to know, and thanks for your feedback!
